source/slang/slang-check-overload.cpp

```diff
@@ -1517,6 +1517,25 @@ namespace Slang
                 return CreateErrorExpr(expr);
         }
 
+        for (auto& arg : expr->arguments)
+        {
+            arg = maybeOpenExistential(arg);
+            // If arg.type is refering to an interface decl with a ThisTypeSubstitution, use the
+            // substituted type.
+            if (auto declRefType = as<DeclRefType>(arg->type))
+            {
+                arg->type->substitute(getASTBuilder(), declRefType->declRef.substitutions);
+                if (auto interfaceRef = declRefType->declRef.as<InterfaceDecl>())
+                {
+                    if (auto thisTypeSubst = findThisTypeSubstitution(
+                            declRefType->declRef.substitutions, interfaceRef.getDecl()))
+                    {
+                        arg->type = thisTypeSubst->witness->sub;
+                    }
+                }
```

Comment:
Basically, this is the logic I expected to see in the constraint inference part of things.

Here in the argument-related stuff I'm not 100% sure it is valid to make this change, because we are overwriting the type of an `Expr` to something other than its original type. Maybe it happens to work out all right in practice, but this makes me nervous, because a type of `IFoo` with a `ThisType` of `Bar` is distinct from the type `Bar` in our system today (although maybe that is actually a broken concept in our representation), and here you are replacing the first type with the latter.